delira/training/base_trainer.py

```diff
@@ -394,83 +467,63 @@ def reduce_fn(batch):
         else:
             raise ValueError("No valid reduce mode given")
 
-        metrics_val = {}
+        for epoch in range(self.start_epoch, num_epochs + 1):
 
-        val_metric_fns = {}
+            self._at_epoch_begin(val_score_key, epoch,
+                                 num_epochs)
 
-        for k, v in self.val_metrics.items():
-            if not k.startswith("val_"):
-                k = "val_" + k
+            batch_gen_train = datamgr_train.get_batchgen(seed=epoch)
 
-            val_metric_fns[k] = v
+            # train single network epoch
+            train_metrics, train_losses = self._train_single_epoch(
+                batch_gen_train, epoch, verbose=verbose)
 
-        if self.metric_keys is None:
-            val_metric_keys = None
+            total_metrics = {
+                **train_metrics,
+                **train_losses}
+
+            # validate network
+            if datamgr_valid is not None and (epoch % self.val_freq == 0):
+                # next must be called here because self.predict_data_mgr
+                # returns a generator (of size 1) and we want to get the
+                # first (and only) item
+                val_metrics = next(
+                    self.predict_data_mgr_cache_metrics_only(
+                        datamgr_valid, datamgr_valid.batch_size,
+                        metrics=self.metrics,
+                        metric_keys=self.metric_keys,
+                        verbose=verbose))
+
+                val_metrics = {"val_" + k: v
+                               for k, v in val_metrics.items()}
+
+                total_metrics.update(val_metrics)
+            _, total_metrics = self._convert_to_npy_fn(**total_metrics)
+
+            for k, v in total_metrics.items():
+                total_metrics[k] = reduce_fn(v)
+
+            # check if metric became better
+            if val_score_key is not None:
+                if val_score_key not in total_metrics:
+                    if "val_" + val_score_key not in total_metrics:
+                        logger.warning(
```

Comment:
Maybe change this to a normal warning (I think the warning should be printed to the terminal independent of the logger configuration)?